File: packages/research_management/src/research_management/models/alert.py
"""
Alert models for monitoring and notifications.
"""

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


class AlertType(str, Enum):
    """Types of alerts in the system."""

    NO_ADVISOR = "no_advisor"  # Student without advisor for > 14 days
    NO_UPDATE = "no_update"  # Project without update for > 30 days
    DEADLINE_SOON = "deadline_soon"  # Deadline approaching in 7 days
    MEETING_REMINDER = "meeting_reminder"  # Meeting in 24 hours


class AlertSeverity(str, Enum):
    """Alert severity levels."""

    INFO = "info"  # 🟢 Informational
    WARNING = "warning"  # 🟡 Requires attention
    CRITICAL = "critical"  # 🔴 Urgent action needed


class AlertStatus(str, Enum):
    """Alert status."""

    ACTIVE = "active"
    RESOLVED = "resolved"
    DISMISSED = "dismissed"


class Alert(BaseModel):
    """Alert model for system notifications."""

    alert_id: str = Field(..., description="Unique alert identifier")
    type: AlertType = Field(..., description="Type of alert")
    project_id: Optional[str] = Field(None, description="Related project ID")
    user_id: Optional[str] = Field(None, description="Related user ID")
    message: str = Field(..., description="Alert message")
    severity: AlertSeverity = Field(..., description="Alert severity level")
    status: AlertStatus = Field(
        default=AlertStatus.ACTIVE, description="Current alert status"
    )
    created_at: datetime = Field(
        default_factory=datetime.utcnow, description="Alert creation timestamp"
    )
    resolved_at: Optional[datetime] = Field(
        None, description="Alert resolution timestamp"
    )

    class Config:
        json_schema_extra = {
            "example": {
                "alert_id": "alert-001",
                "type": "no_advisor",
                "project_id": None,
                "user_id": "user-789",
                "message": "🔴 CRITICAL: Student João Silva is without advisor for 18 days",
                "severity": "critical",
                "status": "active",
                "created_at": "2025-02-01T09:00:00Z",
                "resolved_at": None,
            }
        }


class AlertCreate(BaseModel):
    """Model for creating an alert."""

    type: AlertType
    project_id: Optional[str] = None
    user_id: Optional[str] = None
    message: str = Field(..., min_length=1)
    severity: AlertSeverity

File: packages_dashboard/research_dashboard/lib/src/models/alert.dart
/// Alert model
class Alert {
  final String alertId;
  final AlertType type;
  final String? projectId;
  final String? userId;
  final String message;
  final AlertSeverity severity;
  final AlertStatus status;
  final DateTime createdAt;
  final DateTime? resolvedAt;

  const Alert({
    required this.alertId,
    required this.type,
    this.projectId,
    this.userId,
    required this.message,
    required this.severity,
    required this.status,
    required this.createdAt,
    this.resolvedAt,
  });

  factory Alert.fromJson(Map<String, dynamic> json) {
    return Alert(
      alertId: json['alert_id'] as String,
      type: AlertType.fromString(json['type'] as String),
      projectId: json['project_id'] as String?,
      userId: json['user_id'] as String?,
      message: json['message'] as String,
      severity: AlertSeverity.fromString(json['severity'] as String),
      status: AlertStatus.fromString(json['status'] as String),
      createdAt: DateTime.parse(json['created_at'] as String),
      resolvedAt: json['resolved_at'] != null
          ? DateTime.parse(json['resolved_at'] as String)
          : null,
    );
  }
}

/// Alert type enum
enum AlertType {
  noAdvisor('no_advisor'),
  noUpdate('no_update'),
  deadlineSoon('deadline_soon'),
  meetingReminder('meeting_reminder');

  final String value;
  const AlertType(this.value);

  static AlertType fromString(String value) {
    return AlertType.values.firstWhere((e) => e.value == value);
  }

  String get displayName {
    switch (this) {
      case AlertType.noAdvisor:
        return 'Sem Orientador';
      case AlertType.noUpdate:
        return 'Sem Atualização';
      case AlertType.deadlineSoon:
        return 'Prazo Próximo';
      case AlertType.meetingReminder:
        return 'Lembrete de Reunião';
    }
  }
}

/// Alert severity enum
enum AlertSeverity {
  info('info'),
  warning('warning'),
  critical('critical');

  final String value;
  const AlertSeverity(this.value);

  static AlertSeverity fromString(String value) {
    return AlertSeverity.values.firstWhere((e) => e.value == value);
  }

  String get displayName {
    switch (this) {
      case AlertSeverity.info:
        return 'Info';
      case AlertSeverity.warning:
        return 'Aviso';
      case AlertSeverity.critical:
        return 'Crítico';
    }
  }

  String get emoji {
    switch (this) {
      case AlertSeverity.info:
        return '🟢';
      case AlertSeverity.warning:
        return '🟡';
      case AlertSeverity.critical:
        return '🔴';
    }
  }
}

/// Alert status enum
enum AlertStatus {
  active('active'),
  resolved('resolved'),
  dismissed('dismissed');

  final String value;
  const AlertStatus(this.value);

  static AlertStatus fromString(String value) {
    return AlertStatus.values.firstWhere((e) => e.value == value);
  }
}
